### DISCUSSION TITLE
'wlroots' version

### DISCUSSION THREAD

**Discussion by johanmalm** [labwc/labwc#1178] [Ideas]
Just opening this to allow an ongoing discussions on what 'wlroots' version we track.

@jlindgren90 
@Consolatis 

Follow on from https://github.com/labwc/labwc/pull/1172#issuecomment-1763517762

> Maybe worth its own discussion thread -- how would it be to depend on a soft-fork of wlroots with a select few additional changes (with the intent to get them into mainline wlroots in time)? It could unblock several existing PRs as well as make it possible to add stuff like this _NET_WM_WINDOW_TYPE support. The technical issues seem already mostly solved via meson subproject.

Options:

1. Track latest release (as we currently do with `0.16`). Pros: (a) easiest for users who want to use a packaged release; (b) enables us to ship new features in packaged releases much more frequently than `wlroots` bumps version (which are currently annual); (c) the preference of package maintainers

2. Track `master`. Maintain (fixes only) tagged branch against latest `wlroots` release. Pros: (a)  can implement new features in wlroots and quickly use them; (b) more "continuous" eyes on `wlroots` changes

3. Maintain our own bundled version of `wlroots` with a view to merge with upstream in slow time.

EDIT1: Remove references to `next` branch.

**Comment by Consolatis** [comment]
1. Raises questions on how that `next` branch should work, e.g. is it ok to regularly rebase it on latest master? Manually syncing 0.6 and 0.5 branches was a pain (e.g. backward and forward porting and keeping track of everything). Keep in mind that most issues come in from wlroots 0.16 users and thus get debugged and fixed there first.
2. Not much a fan of that, we were there and did that. It was honestly kind of annoying to keep chasing upstream changes fast because otherwise AUR labwc-git packages (and everybody manually compiling from master) would have their installation break regularly. The second issue is that if development happens on master tracking wlroots master, how would we ever release new features (rather than manually backported fixes) for the 0.6 branch?
3. Absolutely not a fan of that. We should work upstream rather than starting our own fork (be it a soft one or maybe at some point in the future not so soft, e.g. upstream doesn't accept one of our changes but we already rely on it).
4. There is also another variant of 2. that we did before: rather than tracking wlroots master branch, track specific wlroots commits and from time to time update that commit based on the availability of the chase commits. That was slightly less annoying but also causes labwc-git + system wide wlroots-git  to break regularly.

So based on these options I prefer 1. + a manually rebased -next branch that contains the wlroots 0.17 PR + further features / fixes that require wlroots 0.17 to work. It may be some work to keep them rebased on top of master though. I think our priority should be on the master branch (tracking wlroots 0.16) as the majority of users use either that branch directly or a tagged release from it. So shifting the rebase work to the next branch seems sensible to me.

**Comment by jlindgren90** [comment]
(This got really long-winded, sorry...)

In general: a personal view of mine is that it’s good to (1) keep the master branch usable at all times, as much as possible, and (2) release frequently from master. It avoids forcing users to pick between stability or new features and thus helps to reduce fragmentation since everyone is running/testing/developing mostly the same code.

I think we’re struggling with the fact that wlroots does neither of these things. Releases are infrequent (~1 year interval) given the relative immaturity of the library (it might be more excusable for a more mature project) and it’s not uncommon for the master branch to be unusable for weeks at a time between releases.

Regarding the options in particular:

1. I am not sure I understand what’s being described here. We would have two different “current” versions of labwc that distributions could pick from, using wlroots latest release vs. wlroots master branch? If that’s the idea, I honestly don’t see how it benefits anyone. It’s more work for us, confusing for distribution maintainers, and leaves users stuck having to choose between features vs. stability.

2. (Labwc master tracks wlroots master, while a separate stable branch uses wlroots latest release.) I want to like this idea; it seems sensible in theory. Unfortunately it doesn’t really do much to address the issues caused by wlroots. The stable branch would fall too far behind on features, while the master branch would be too unstable for daily use.

3. This was my proposal: “soft-fork” wlroots with a branch based on wlroots latest release, plus cherry-picked commits from wlroots master or open wlroots merge requests. I think this is the best option for users since they get all of our latest features without the instability of wlroots master. The downsides I see are mainly for distributors (all the same endlessly rehashed objections to packaging bundled libraries, that don’t need to be repeated here) and for us (a bit more work since it’s on us to stay on top of rebasing to each wlroots point release).

Regarding the concern of the soft-fork eventually diverging from upstream wlroots: I would absolutely want to prevent that. If a wlroots MR was rejected upstream, I would say we should revert it in our branch/fork, and find some other way to implement the feature/solve the problem. And our branch/fork ought to be rebased on each new stable wlroots release (though we could keep old branches around to preserve history if desired).

If option 3 is still undesirable, then my second choice would basically be the status quo (labwc master tracks wlroots latest release, labwc releases are from master).

I think the long-running and continually rebased wlroots-0.17 branch/pull-request is useful for us to keep up with wlroots development and not have to play catch-up all of a sudden when wlroots 0.17 is released. I don’t think it’s really useful for most users (not stable enough) and I like the current policy of keeping its changes minimal (i.e. no new features/unnecessary refactoring). It doesn’t really matter to me whether that branch is hosted under the labwc project or continues to live on my personal GitHub.

Separately, I recently created a 0.16-backports wlroots branch (also on personal GitHub) which is essentially the “soft-fork” I proposed in option 3. In my view this wlroots branch has a different goal than the wlroots-0.17 labwc branch: enabling new labwc features vs. keeping up with wlroots changes. I want this branch to be stable enough for daily use, whereas the wlroots-0.17 branch is not.

Since I am already effectively maintaining a fork of labwc for personal use, I would love it if other users could benefit from that work. I certainly won’t be offended if that 0.16-backports approach is not seen as desirable for labwc upstream, and in that case I can just host a “labwc with patches” branch in my personal GitHub.

Anyway, thanks for the discussion, I enjoy/appreciate it regardless of the outcome :)

**Reply by Consolatis** [reply]
> (This got really long-winded, sorry...)

Absolutely welcome from my side. It helped clean up a misunderstanding if mine of your proposal from a soft fork: I was under the impression that you were referring to a  constantly rebased version of wlroots 0.17 with additional patches.

It didn't really change my stance on the issue though: it breaks all package updates (not just packaging). So when we were to do that, most likely none of the labwc packages would get any more updates unless the distro is fine to build our wlroots fork statically. This would affect people that may not even know that this git repo exists because they just found an "Openbox inspired wayland compositor" in their repo and gave it a try.

This issue could likely be worked around with some meson magic that sets a define in config.h if built with our fork. That define could then guard all functionality within labwc that requires the fork. But this feels *very* hacky, clutters the codebase and will likely be hard to maintain long term (for example what should happen when this feature requires either our 0.16 fork or wlroots >= 0.17).


> 1. I am not sure I understand what’s being described here. We would have two different “current” versions of labwc that distributions could pick from, using wlroots latest release vs. wlroots master branch? If that’s the idea, I honestly don’t see how it benefits anyone. It’s more work for us, confusing for distribution maintainers, and leaves users stuck having to choose between features vs. stability.

I don't think any distribution (apart from maybe the AUR -git package) would want to deal with a moving wlroots. So I don't think that this would confuse distribution maintainers. For them, nothing changes. Usual users should also keep using the latest release or the master branch if they feel like it.

There are some users that require wlroots 0.17 for other reason like some high-dpi stuff or other new protocols added to 0.17. For these users instead of having to clone the master repo + applying the 0.17 patchset + potential further patchsets it would make life easier because they now just need to use some `0.7-dev` branch.

It would also allow to merge the pending PRs that target features in wlroots 0.17. This indeed contradicts the other goal of keeping the 0.17 patchset as small as possible but who knows when wlroots actually does a release. There is some other stuff [piled up](https://github.com/labwc/labwc/issues/727) that has to / should be implemented once we switch so having the ability to work on that before a wlroots release is also a benefit.

@heroin-moose for example spoke about implementing fractional scale support in labwc which requires wlroots 0.17.
Having a branch to target would make these kind of things much easier.

> Anyway, thanks for the discussion, I enjoy/appreciate it regardless of the outcome :)

Yes, I do too :)

**Comment by jlindgren90** [comment]
> most likely none of the labwc packages would get any more updates unless the distro is fine to build our wlroots fork statically

I am not too up-to-date on the various distros' policies on static libraries. I've never worked in that role. My experience maintaining Audacious over the past 10-15 years was that distros didn't really care what we bundled/unbundled. Though in those cases we were copying code into our Git repo/release tarballs, not referencing an external Git repo like meson subprojects do. Maybe that makes a difference.

Let's say we continue the current approach for the master branch. Do as much development as we can there, given the latest wlroots release, and tag our official releases from master. Then I guess we have several possibilities for unofficial/semi-official "next" branches that could be periodically rebased on master:

- minimal wlroots master/0.17 compatibility fixes, no additional changes (like my current wlroots-0.17)
- same plus new feature development requiring wlroots master/0.17 (like the "next" branch idea mentioned previously)
- new feature development but targeting a stable-ish wlroots 0.16 + backports branch (my "soft fork" idea)

Those aren't mutually exclusive, we could maintain any/all of those given the time and interest. Naming (of course) and what/how to recommend to users would be something to figure out.

**Reply by johanmalm** [reply]
Oh I like this a lot.
Sorry, my language was a bit clumsy in OP (e.g. use of `next`).
I wonder if there would be much appetite in the `wlroots` team to release more often - like the Linux kernel have ended up doing.
I might just post something on IRC to see.

**Comment by jlindgren90** [comment]
So here is a proof-of-concept "new feature development but targeting a stable-ish wlroots 0.16 + backports" branch: https://github.com/labwc/labwc/pull/1180

Not sure if a pull request is the appropriate format/presentation, but it works for now.

**Comment by heroin-moose** [comment]
Actually, there is a third option. We can try to bring wlroots and  few other parties like sway, hyprland, etc and discuss a possibility of more granular wlroots releases (like one in three months or so). I think this will benefit most of the wlroots users.

**Reply by johanmalm** [reply]
For those not on IRC. Discussion ongoing on `#wlroots`

**Reply by jlindgren90** [reply]
I apologize for not being able to join in on IRC, but I would note that more frequent releases alone won't help much unless they are also more stable than wlroots master currently is, and unless open MRs are also reviewed in a more timely fashion. I don't mean to criticize the wlroots project here, and they are doing good work, it just seems like issues stemming from a lack of manpower and/or organization. I'd be glad to know if there's some way I can help.

**Comment by Consolatis** [comment]
While emersion prefers compositors to always track wlroots master (to get more eyes on the wlroots code) I think we mostly agree (correct me if wrong) on keep developing based on the last wlroots release via our master branch (or rather based on the wlroots branch of the last release which includes all commits making up the point releases but provides a frozen API).

IMHO stability for us is way more important than riding the `look: a new feature` wave. In my eyes, a compositor serves as the base to run other applications and thus should be as stable as possible, even when using the `master` branch. This also increases the amount of people running it and reporting bugs / regressions.

Sorry, the following is a bit of an unordered brain dump:

So the question is what additional branches we should host / maintain in the labwc repo. My preference would be
- `0.7-dev` which
  - gets rebased regularly on our master branch
  - includes the wlroots 0.17 PR + pending PRs targeting 0.17
  - serves as a target branch for future PRs that require 0.17
- I would be fine to allocate some time to keep this branch rebased on top of master but help here would be much appreciated
- when wlroots 0.17 gets released:
  - we update the `0.7-dev` branch `wlroots.wrap` file to track the `0.17` branch rather than `master` (not just the 0.17 release tag)
  - when we decide to switch our master branch over to 0.17 (which IMHO should also take the distro packaging status of wlroots 0.17 into account):
    - we create a new `0.6` branch based on current `master` like we did for `0.5`
      - the `0.6` branch from that point on is only receiving bug fixes but no new features
      - we may do new releases from the `0.6` branch if the severity of the fixes call for it
      - the `master` branch is frozen
    - we create a PR to get the commits from `0.7-dev` into `master`
      - we also have a chance to clean them up some more, e.g.
        - remove some commits and fix others that are currently somewhat broken (nested instance resizing mainly)
    - we accept the PR
    - we open up `master` again, now tracking the wlroots `0.17` branch
    - we retarget existing PRs against `0.7-dev` to `master`
    - we delete `0.7-dev`
    - after some testing we release labwc `0.7.0` based on `master`
    - at this point distros can choose if they want to stick to the `0.6.x` releases (e.g. they are still on wlroots 0.16) or use the `0.7.x` releases.
    - we could potentially create a new `0.8-dev` branch here and start back at the top with `s/0.17/0.18/; s/0.7/0.8/`.

Regarding the soft fork variant (I'll call it `wlroots-0.16-backports` here but obviously that name could be changed):
I'd be fine if we host it in the labwc repo but would likely not spend much time on rebasing work here.
One question I have regarding this branch is about its lifetime, e.g. what happens when master starts to track wlroots 0.17.

> minimal wlroots master/0.17 compatibility fixes, no additional changes (like my current wlroots-0.17)

I am not sure we actually need this anymore when we have a `0.7-dev` branch. That one should serve the same end goal.
Or do I miss a benefit of keeping that around?

So with these two new branches some questions arise regarding the target branches for new features / fixes:
- Which branch should we target? My preference here would be
  - First and foremost the `master` branch if at all possible
  - If a feature does require changes in wlroots
    - they are already merged upstream:
      - `0.7-dev` + eventually cherry-picked / backported into `wlroots-0.16-backports`
    - they are *not* yet merged upstream:
      - `wlroots-0.16-backports` with a requirement to also send a PR for `0.7-dev`

These are just my current thoughts on how we could make this work with as little friction as possible, any further suggestions, criticism or other opinions are obviously very welcome.

**Comment by jlindgren90** [comment]
For wlroots-0.16-backports:

- I would suggest each change/feature included in this branch should still have a separate open draft PR (with the depends-on-wlroots-change tag) which merges cleanly to master. The changes would then be cherry-picked to the wlroots-0.16-backports branch, but the PRs would remain open for later review/merge to master once master targets the needed wlroots version.

- If there is enough of a review process around 0.7-dev (more on that below), and the needed wlroots changes aren't difficult to backport, then changes could also be cherry-picked from 0.7-dev to wlroots-0.16-backports (no need to keep a separate PR open).

- Once labwc master has all those PRs merged, the wlroots-0.16-backports branch would go away/be archived. If there are wlroots changes we need (with open wlroots MRs) that didn't make it into 0.17, then we could create a wlroots-0.17-backports branch instead.

I'm happy to take care of curating this branch if there are no objections.

For 0.7-dev, my main question is when/how code reviews happen. Do we have PRs targeting this branch, which are reviewed/merged? Do they have to be reviewed a second time before being merged to master (when master switches to wlroots 0.17)?

The current wlroots-0.17 branch is kind of a "wild west" as to pushing/rebasing/force-pushing without any review. That's been okay because of the narrow scope of the changes, but if 0.7-dev is going to have more development going on (and more people working on it) then IMHO it would be nice to have a little more process/rules around it.

Brainstorming a bit:
- Could we have 0.7-dev normally be under PR control: new development and new compatibility fixes for breaking wlroots changes need PRs to merge to 0.7-dev?
- Simple merges from master (no force-push, no significant merge conflicts) could be allowed without PRs.
- Rebasing onto master (which requires force-push) could be done less frequently, with perhaps a rule that the end result needs to exactly match the current HEAD of the 0.17 branch?

**Comment by Consolatis** [comment]
> I'm happy to take care of curating this branch if there are no objections.

Nope, non at all from my side.

> For 0.7-dev, my main question is when/how code reviews happen. Do we have PRs targeting this branch, which are reviewed/merged? 

Yes, I think we should apply the same rules as for the master branch (besides the rebasing on master itself).
Also, PRs which could also target `master` should be directed there instead and then end up in the `0.7-dev` branch via the next rebase.
We could for example start with retargeting the strut_partial PR to `0.7-dev`.

> Do they have to be reviewed a second time before being merged to master (when master switches to wlroots 0.17)?

No, although we could have a final review of all the accumulated commits if we want to.

> The current wlroots-0.17 branch is kind of a "wild west" as to pushing/rebasing/force-pushing without any review. That's been okay because of the narrow scope of the changes, but if 0.7-dev is going to have more development going on (and more people working on it) then IMHO it would be nice to have a little more process/rules around it.

Yes, completely agree.

> Could we have 0.7-dev normally be under PR control: new development and new compatibility fixes for breaking wlroots changes need PRs to merge to 0.7-dev?

See above: I'd be all for that.

> Simple merges from master (no force-push, no significant merge conflicts) could be allowed without PRs.

Not sure I follow here, With merges (especially with merge conflict resolutions) wouldn't we diverge from `master` and thus make it harder to regularly rebase onto master + merge back into master when switching it to 0.17?

> Rebasing onto master (which requires force-push) could be done less frequently, with perhaps a rule that the end result needs to exactly match the current HEAD of the 0.17 branch?

Care to explain the last part a bit more? My recollection of how rebases work is that all commits after a shared ancestor from the branch being rebased have a different commit hash (unless the rebase is a no-op). But I may be wrong here.


Another idea that came up: we could start the `0.7-dev` branch as a copy of `master` and then retarget the #626 PR to it.
That would allow us to review the chase commits once more before we settle on them for the foreseeable future.

**Comment by jlindgren90** [comment]
Maybe merges from master would be wasted work and it would be better to just rebase regularly, I don't know.

The advantage with merges, in my mind, is for preserving history and review, because you can easily see how conflicts were resolved (git show on the merge commit). Often I will do a merge first, make sure everything is working, and only then rebase, making sure that the git diff from the merged version to the rebased version is empty. That's what I meant by the end result exactly matching. Commit history is different but file contents of HEAD are the same.

When you rebase of course the merge commit is omitted, all the conflict resolutions have to be applied to each rewritten commit instead. A good difftool helps a lot there as then you can just copy the resolutions from the merged version rather than redoing it from scratch.

I had it in my head that the normal/more frequent process for keeping 0.17-dev up to date with master would be merging, because there's less risk of accidental destruction (omitting a commit by mistake etc.) Then maybe we would have someone give advance notice when they were planning to do a rebase, with maybe a few days to let things settle and make sure the last merged state was good first.

I don't really know the best way to manage this with a collaborative branch. Just thinking out loud, I guess.

**Comment by jlindgren90** [comment]
Definitely agree that PRs should target master first and only 0.7-dev if necessary. And PRs would not target the backports branch directly, that would be more like a downstream of the other two, and in limited cases of open PRs targeting them.

Starting 0.7-dev as a copy of master and retargeting #626 to it sounds good.

**Comment by jlindgren90** [comment]
A specific example of where rebase has been painful with the wlroots-0.17 branch is the try_from introduction. That "chase" commit has been modified/expanded multiple times as a result of changes in master that add new "from" calls needing to be changed to "try_from". I don't know how to manage that in a collaborative branch. Making those changes silently as part of a rebase, without any review, seems dirty/risky.

**Reply by Consolatis** [reply]
What I sometimes do in my PR branches in those cases is adding a new commit like "[fixup] Chase wlroots: convert to try_from".
Then on the next rebase I move the commit after the original one (identified by whatever is listed after the [fixup] tag) to see if there are any follow-up conflicts. If not, I'll rebase once more and mark all [fixup] commits as "f" thus apply the fixups.

I am not sure if that is the best way to do it but it served me well. But those branches were always under my full control so I am not sure how that works in a shared branch. In either case it would allow to review PRs with such changes and delays potential follow-up conflicts until the next rebase.

**Comment by jlindgren90** [comment]
More thinking out loud:

- It would be nice to have the meson subproject in 0.7-dev reference a commit hash rather than just "master". Helpful to make sure things build correctly while rebasing, and extremely helpful if we ever need to do a bisect.

- Maybe worth considering just doing merges for the lifetime of the 0.7-dev branch, and do one final rebase before it is merged to/becomes master? Upside is it preserves history & works well with multiple developers pulling/pushing. Downside is it's a bunch of work at all once at the end, but I believe I could manage it.

**Reply by Consolatis** [reply]
> It would be nice to have the meson subproject in 0.7-dev reference a commit hash rather than just "master". Helpful to make sure things build correctly while rebasing, and extremely helpful if we ever need to do a bisect.

We did that before the last wlroots 0.16 release. Its a bit nicer than to always have stuff fail to build due to API changes. Downside of that is that it takes longer to detect any breaking changes. I would be fine with this.

> Maybe worth considering just doing merges for the lifetime of the 0.7-dev branch, and do one final rebase before it is merged to/becomes master? Upside is it preserves history & works well with multiple developers pulling/pushing. Downside is it's a bunch of work at all once at the end, but I believe I could manage it.

Hm.. Trying to think of negative aspects of this approach. Only thing that comes to mind would be cluttered history but that could be corrected on the final merge to master. Keeping track of the merge conflict resolutions to the `0.7-dev` commits indeed sounds like a lot of work though. Maybe they could be stored along the merge message?

**Reply by jlindgren90** [reply]
> Maybe they could be stored along the merge message?

I like this idea. The default "git commit" content usually has these listed (at least the ones it detected) as a "Conflicts:" comment (lines start with #). We could uncomment that for a start.

**Reply by jlindgren90** [reply]
A variation on your "fixup" approach could still be useful too: do the merge, fix the textual conflicts, commit. Then if there are additional "fixups" to apply (like from -> try_from) make those in a separate commit after the merge.

**Reply by Consolatis** [reply]
> A variation on your "fixup" approach could still be useful too: do the merge, fix the textual conflicts, commit. Then if there are additional "fixups" to apply (like from -> try_from) make those in a separate commit after the merge.

And then send a PR for the whole thing. The merge message + `[fixup]` commits would then help on the final rebase before merging to master. Yeah.. I think I like this. It indeed makes stuff way easier to work with if we don't do any rebases on the `0.7-dev` branch at all. E.g. a simple `git pull` just works (tm).

**Comment by jlindgren90** [comment]
Trying to summarize the ideas for `v0.7-dev` so far & fill in some details ...

Briefly:
- v0.7-dev will be branched starting from master.
- #626 will be retargeted to v0.7-dev as the first PR.
- New development that requires wlroots 0.17 should be submitted as PRs targeting v0.7-dev.
- Development not requiring wlroots 0.17 should still target master.

Merging from master:
- Periodically (maybe weekly, give or take?) a Maintainer will merge master into v0.7-dev.
- Any "Conflicts:" section in the merge commit message should be uncommented (this will be helpful for rebasing later).
- Any fixes needed beyond simple conflict resolution should be in a separate "[fixup]" commit.
- The result can be pushed directly (no force-push though), no PR needed.

For new breaking wlroots changes:
- Maybe new "chase wlroots" commits can also be pushed directly by a Maintainer?
- But if it's an invasive fix and the author wants more eyes on it, PR is fine too.
- We should continue to reference the wlroots commit hash in the commit message.
- wlroots.wrap should also be updated to "pin" the subproject to that commit.

When we're ready to switch master over to wlroots 0.17:
- Beforehand, the current state of master would be branched/archived to a v0.6 branch.
- One of us will create a temporary branch from v0.7-dev.
- This temporary branch will be rebased onto master (killing all the merge commits).
- Any "[fixup]" commits would be squashed with related earlier "chase" commits.
- Then a PR targeting master would be created from this branch.

Is that accurate? And @johanmalm has been quiet for a little bit, what are your thoughts?

**Reply by Consolatis** [reply]
I think that sums it up pretty accurately. Only thing missing might be the requirement to always target fixes / features at the master branch (if possible).

I am not sure about the direct pushes, another option would be always making an informal PR and then immediately self-accept (when the "Arch" CI build went through). With this there is a) a GH notification, b) a CI check and c) a place to discuss things after the fact in case something went slightly wrong or there is a follow-up fix required.

**Reply by jlindgren90** [reply]
The only catch with using PRs to merge from master is that we need to be careful not to rebase the commits from master onto v0.7-dev.

**Reply by Consolatis** [reply]
I assume you are talking about the 3 different "accept PR" options on Github? E.g. "Rebase and Merge" vs "Create a merge commit" vs "Squash and merge"? What would be the issue with "Rebase and Merge" here?

**Reply by jlindgren90** [reply]
Rebase and merge would change the commit hash of commits already in master. So next merge they would (potentially, unless GH is smarter than I give it credit for) show up as needing to merged again.

**Reply by jlindgren90** [reply]
If "rebase and merge" preserves the merge commit then we're fine. I assumed it didn't but have not tested it.

**Reply by Consolatis** [reply]
Maybe we could create a temporary test branch and run some PRs against it to ensure that it won't cause issues.
If it indeed causes the 2nd PR to also consider the already existing commits from the merged first PR for inclusion and thus starts to complain about non-existing conflicts that would indeed be pretty bad.

**Reply by jlindgren90** [reply]
https://github.com/jlindgren90/labwc/pull/1

It seems GitHub will not allow rebase-and-merge ("This branch cannot be rebased due to conflicts") when there are any merge conflicts, even if resolved by a merge commit included in the PR.

**Reply by jlindgren90** [reply]
Added "Development not requiring wlroots 0.17 should still target master" to the summary for clarity.

**Reply by johanmalm** [reply]
Managed to read through to the end eventually :smile:

I think it all looks very sensible and think we should go with what you've summarized.

Just to confirm I've understood it correctly, will `0.7-dev` contain new `0.17` dependent features too (rather than open PRs).

I like the approach of merging when chasing `wlroots`, and only rebasing occasionally. It gives us much more control and confidence. With [fixup] commits, I assume we need to specify which commit the fixup relates to to make the interactive-rebase easier later.

> Maybe new "chase wlroots" commits can also be pushed directly by a Maintainer?

Yes, I think so.

> Often I will do a merge first, make sure everything is working, and only then rebase, making sure that the git diff from the merged version to the rebased version is empty. 

I like that.

> When you rebase of course the merge commit is omitted, all the conflict resolutions have to be applied to each rewritten commit instead. A good difftool helps a lot there as then you can just copy the resolutions from the merged version rather than redoing it from scratch.

I don't quite follow this and probably need some education. I just rebase from the command line, but yeah it's quite hard sometimes. What do you use... `kdiff3`?

**Reply by jlindgren90** [reply]
> Just to confirm I've understood it correctly, will 0.7-dev contain new 0.17 dependent features too (rather than open PRs).

Yes :+1: 

> I like the approach of merging when chasing wlroots, and only rebasing occasionally.

Consolatis and I had discussed rebasing only once at the end, but if we want to do it occasionally in between, that's okay with me too. It would be nice to give others some notice if one of us is planning to do that.

> I don't quite follow this and probably need some education. I just rebase from the command line, but yeah it's quite hard sometimes. What do you use... kdiff3?

I have git set-up to use Beyond Compare (proprietary but Linux-native) for merge conflicts and for `git difftool`. I used it at a previous job and haven't found anything open-source that I like as well, so I gave in and paid for a license. Not to sound like an ad or anything, but I do find it invaluable for merging/rebasing. One time I had to merge a Linux stable branch (4.4.x) back into a vendor kernel that had diverged and been abandoned years earlier; I can't imagine having done it if I had to manually edit merge conflict sections.

Meld and kdiff3 are alternatives, but Meld cannot really do a 3-way merge, and overall I find just kdiff3 buggy/clunky.

**Comment by johanmalm** [comment]
> Is that accurate? And @johanmalm has been quiet for a little bit, what are your thoughts?

Rest assured no response ≠ no interest :smile: Just have a busy day-job.

I'm not very good with this type of process stuff (through lack of experience). Will read through over weekend and try to take it in.

**Comment by johanmalm** [comment]
@jlindgren90 @Consolatis 
So, `0.6.6` is out. It feels to me like we should now:
1. Prioritize porting to `wlroots 0.17` on `master`. When ported and we've had a period to grow in confidence that we've got no regressions, we release `0.7.0`
2. If/when there are fixes on the `0.6.X` line, but create an `0.6` branch on push them there.
Thoughts?

**Reply by jlindgren90** [reply]
Sounds good to me.

**Reply by johanmalm** [reply]
So how about proceeding as follows:

1. Stop merging anything else on `master`
2. Rebase https://github.com/jlindgren90/labwc/tree/wlroots-0.17
3. Tidy up `[fixup]` etc commits
4. Review branch above
5. Merge

**Reply by johanmalm** [reply]
Have reviewed. Looks good. 

Comments:

1. Set `revision = 0.17` in `subprojects/wlroots.wrap`
2. The `try_from` stuff looks deep, but happy with `chase`
3. Do we leave `subprojects/libdisplay-info.wrap` as is?

Looks like are all "post-0.17"
https://github.com/swaywm/sway/commit/47e6a1164c25b5b49bfb919a681b88641d2ce37c
https://github.com/swaywm/sway/commit/128b6253a924c30e8e4fd1f09ce19e21a826c5ad
https://github.com/swaywm/sway/commit/fd6d6f1d97deea2fc94aaeb58c85c99a498d905b
https://github.com/swaywm/sway/commit/a946b1aecfa3f32a3bfae66d66df0bdf77440d69


Sidenote: I wonder if we're missing some stuff in output.c wrt https://github.com/swaywm/sway/commit/4326a26ad6fb64b8f8de7d32c1e7b6423e1f84f3

**Reply by johanmalm** [reply]
@jlindgren90 

Looks like the following is needed to get running:

```
diff --git a/meson.build b/meson.build
index bf98578..b95587f 100644
--- a/meson.build
+++ b/meson.build
@@ -68,6 +68,7 @@ glib = dependency('glib-2.0')
 cairo = dependency('cairo')
 pangocairo = dependency('pangocairo')
 input = dependency('libinput', version: '>=1.14')
+pixman = dependency('pixman-1')
 math = cc.find_library('m')
 png = dependency('libpng')
 svg = dependency('librsvg-2.0', version: '>=2.46', required: false)
@@ -111,6 +112,7 @@ labwc_deps = [
   drm,
   pangocairo,
   input,
+  pixman,
   math,
   png,
 ]
```

**Reply by jlindgren90** [reply]
I am not sure if I will be able to do the rebase today, please feel free to push/force-push to that branch though.

**Reply by johanmalm** [reply]
Okay - no rush.
I've just done a rebase.... and it was clean :smile:
Will push it.

**Comment by stefonarch** [comment]
Here on arch today:

>looking for conflicting packages...
error: failed to prepare transaction (could not satisfy dependencies)
:: installing wlroots (0.17.0-1) breaks dependency 'wlroots<0.17' required by labwc-git

**Reply by johanmalm** [reply]
Arch now has two packages: wlroots and wlroots0.16

@narrat could you help adjust the PKGBUILDs. 
We’re still against 0.16 on everything, but will move master to track 0.17 shortly. Will let you know when we do.

**Reply by stefonarch** [reply]
> Arch now has two packages: wlroots and wlroots0.16

It doesn't propose it here (yet?):

```
$ yay -S wlroots
wlroots                      wlroots-docs                 wlroots-hidpi-xprop-git      wlroots-nvidia-git           
wlroots0.15                  wlroots-full-git             wlroots-no-axrgb-assert-git  wlroots-xrgb8888-git         
wlroots-asan-git             wlroots-git                  wlroots-nvidia               
```
No big issue, I just blocked the package atm.

**Reply by johanmalm** [reply]
So you need to do a `-Sy`?

```
sudo pacman -Sy
pacman -Ss wlroots | grep ^extra/wlr
extra/wlroots 0.17.0-1 [installed]
extra/wlroots0.16 0.16.2-2 [installed]
```

**Reply by Narrat** [reply]
Or the mirror has some problems.
The labwc-git PKGBUILD got adjusted three days ago. So @stefonarch, you need either to refresh your local copy of the PKGBUILD via `git` or download it anew.

**Reply by stefonarch** [reply]
Looks like an issue with `yay` and extra, `pacman` shows 
```
$ pacman -S wlroots
wlroots      wlroots0.16 
```
